[Login] Allow login for WPCOM suspended sites

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14196
Closes: #14255
Closes: #14256
Part of: #14384

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

⚠️ Don't merge until https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/858 is merged.

Adds code to allow login for WPCOM suspended sites. This logic is added behind a configuration flag in the authenticator library. 


## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Prerequisites
- Create a JN site with Woo and Jetpack
- Connect Jetpack from `wp-admin` page
- Follow the steps to mark it as suspended in WPCOM. Internal - p1729584367490299/1729584240.840689-slack-C03L1NF1EA3
- Reach out to me or @hichamboushaba if you need help suspending a site.

Login into a suspended site
- Build and run the app
- Logout if needed
- Login into the app using the JN site
- Check Xcode logs and you should see the following event being tracked.
```
Tracked black_flagged_website_detected, properties: [event: login]
```
- You should be able to login into the app using .org site credentials
- Validate that you don't see the Jetpack banner in the dashboard.

WPCOM credentials login
- Follow the steps and mark the site as not suspended. (Remove flags) Internal - p1729584367490299/1729584240.840689-slack-C03L1NF1EA3
- Logout from the app.
- Try logging in using the same JN site.
- You should be navigated to WPCOM login flow and validate that WPCOM login works.

JN site .org credentials login
- Create a JN site with Woo and no Jetpack
- Logout from the app
- Login into the app using .org site credentials
- Validate that you see the Jetpack banner in the dashboard


## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I tested the following scenarios.

- Suspended site's .org site credentials login works. The Jetpack banner is hidden. 
- Normal site credentials login works as before. The Jetpack banner is displayed for eligible sites. 
- Normal WPCOM credentials login works as before.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

#### Suspended site
| Before | Now |
|--------|--------|
| ![Simulator Screenshot - iPhone 15 Pro Max - 2024-10-30 at 13 45 21](https://github.com/user-attachments/assets/20151e72-6428-4fa0-b54f-a9357359b16f) | ![Simulator Screenshot - iPhone 15 Pro Max - 2024-10-30 at 13 52 15](https://github.com/user-attachments/assets/f536e32f-ba62-4bca-8402-626802e04fbf) |

#### Dashboard - Jetpack banner of site without Jetpack

| Normal JN site (without Jetpack) | Suspended JN site  |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/faa9fedc-7e5d-40ef-a1ef-4461b858556f) | ![image](https://github.com/user-attachments/assets/b4e451a8-4f2d-48cf-b799-307766336e3a) | 


#### Settings - Install Jetpack setting

| Normal JN site (without Jetpack) | Suspended JN site  |
|--------|--------|
| ![Simulator Screenshot - iPhone 15 Pro Max - 2024-11-12 at 14 24 51](https://github.com/user-attachments/assets/2dbba55a-d6f1-46bb-9e7c-fb8f70f1cacd) | ![Simulator Screenshot - iPhone 15 Pro Max - 2024-11-12 at 14 23 43](https://github.com/user-attachments/assets/14f120b5-2a1e-428a-9b94-259bd7a8c272) | 


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

selanthiraiyan committed:
Point authenticator library to commit.

selanthiraiyan committed:
Add user defaults key to store WPCOM site suspended state.

selanthiraiyan committed:
Helper methods to read and write to user defaults.

selanthiraiyan committed:
Inject user defaults.

selanthiraiyan committed:
Update user defaults if authenticator library sends suspended site error.

selanthiraiyan committed:
Clear wpcomSiteSuspended user defaults value upon logout.

selanthiraiyan committed:
Stop showing Jetpack banner for suspended sites.

selanthiraiyan committed:
Set WPCOM suspended site login config as false.

selanthiraiyan committed:
Merge branch 'trunk' into feat/14196-parse-site-info

selanthiraiyan committed:
Point to authenticator library PRs latest commit.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan committed:
Merge branch 'trunk' into feat/14196-parse-site-info

hichamboushaba commented on WooCommerce/Classes/Extensions/UserDefaults+Woo.swift:
```diff
@@ -65,6 +65,10 @@ extension UserDefaults {
         // Selected campaign objective saved for future campaigns
         case blazeSelectedCampaignObjective
+
+        // Whether the site is suspended on WordPress.com and can't be connected using Jetpack
+        //
+        case wpcomSiteSuspended
```
Just a question for my learning, do we have any guidelines on where to use `UserDefaults` vs something like `GeneralAppSettings` in the app? in the custom fields project, I used `GeneralAppSettings`, and it needed a lot of boilerplate to my liking, so I'm not sure if I should've used `UserDefaults` instead 😅

selanthiraiyan replied:
> do we have any guidelines on where to use UserDefaults vs something like GeneralAppSettings in the app?

Good question. I don't think there are any guidelines. 

Personally, I prefer user defaults when I have to update UI when the values change. Because, we can do [this](https://github.com/woocommerce/woocommerce-ios/pull/14257/files#diff-5e4b4001de415046921cf84e4ce2b6c16646ae1bb2067aa6d9ec1fbfa2a855e5R499-R504) to listen to the value using Combine and update UI.

hichamboushaba commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:
```diff
@@ -60,6 +60,8 @@ final class DashboardViewModel: ObservableObject {
 
     @Published private(set) var jetpackBannerVisibleFromAppSettings = false
 
+    @Published private(set) var isSiteEligibleToInstallJetpack = true
```
We need a similar logic for the app settings to hide the "Install Jetpack" button.

selanthiraiyan replied:
Great catch. Done in 59a6121ff0e211d38068c683e5749c01ec07b8f2. Thank you!

hichamboushaba reviewed (commented):
Nice work @selanthiraiyan, the logic works well.

There are just two points that needs addressing before approving the PR, the unit tests are failing, and the "Install jetpack" button is still shown in the app settings.

selanthiraiyan committed:
Merge branch 'trunk' into feat/14196-parse-site-info

selanthiraiyan committed:
Fix error in unit tests.

selanthiraiyan committed:
Hide Install Jetpack in settings for WPCOM suspended site.

selanthiraiyan committed:
Add tracking events and keys for WPCOM suspended site case.

selanthiraiyan committed:
Track event when WPCOM suspended site is detected during login.

selanthiraiyan committed:
Enable the configuration to allow login into WPCOM suspended sites.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan commented:
Thank you for the review, @hichamboushaba! 

I addressed your comments and added tracking. I have also enabled this feature and updated the PR description accordingly. 

The PR is ready for another look. 🙏

hichamboushaba reviewed (approved):
Nice work @selanthiraiyan, this works well.

Just a **non blocking** nitpick, I'm aware most of us don't care about a good git history, but personally I do, and find it quite useful when debugging something, so I would suggest to write this branch's history by squashing the last commits that were about fixing the unit tests.

selanthiraiyan committed:
Revert changes to cocoapods lockfile.

selanthiraiyan committed:
Merge branch 'trunk' into feat/14196-parse-site-info

selanthiraiyan committed:
Point authenticator to commit from trunk

selanthiraiyan commented:
> I would suggest to write this branch's history by squashing the last commits that were about fixing the unit tests.

I agree. Done.
